File: 714-best-time-to-buy-and-sell-stock-with-transaction-fee/best-time-to-buy-and-sell-stock-with-transaction-fee.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices, int fee) {
        vector<vector<int>>dp(prices.size() + 1, vector<int>(2, -1));
       return solve(0, true, fee, prices, dp); 
    }

    int solve(int i, bool buy, int &fee, vector<int>& prices, vector<vector<int>>&dp){
        if(i >= prices.size()) return 0;
        if(dp[i][buy] != -1) return dp[i][buy];
        if(buy){
            int take = - prices[i] + solve(i + 1, false, fee, prices, dp);
            int nonTake = solve(i +1, true, fee, prices, dp);
            return dp[i][buy] =  max(take, nonTake);
        }
        int sell = prices[i] - fee + solve(i + 1, true, fee, prices, dp);
        int nonSell = solve(i + 1, false, fee, prices, dp);

        return dp[i][buy] = max(sell, nonSell);
    }
};

File: 714-best-time-to-buy-and-sell-stock-with-transaction-fee/best-time-to-buy-and-sell-stock-with-transaction-fee.py
class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        # You initially have zero money in your hand
        cin_w_shares = -prices[0] # Cash in hand with shares if you bought it on first day
        cin_wo_shares = 0 # Cash in hand without shares if you do not buy any share on first day
        
        for i in range(1,len(prices)):
            # Maximum cash in hand with shares
            # Either 
			# 1. withold prev share in which case your cash in hand will not change, 
            # 2. or assume there was no currently bought share but you want to buy it today - 
			#         In this case: your current cash in hand with shares will be your previous cash 
			#         in hand without shares minus buying price of the share today.
            cin_w_shares = max(cin_w_shares, cin_wo_shares-prices[i]) 
            
            # Maximum cash in hand without shares
            # Either 
			# 1. withold money without shares in which case your cash in hand will not change, 
            # 2. or assume you previously bought the share and you are going to sell that today -
			#         In this case : your current cash in hand without shares will be your previous cash 
			#         in hand with shares plus the current selling price minus transaction fee
            cin_wo_shares = max(cin_wo_shares, cin_w_shares + prices[i]-fee)
            
        # Return cash in hand without shares, as cash in hand with share will always be lower
        return cin_wo_shares 